app/code/Magento/Catalog/Model/Product/Option.php

```diff
@@ -306,44 +323,30 @@ public function setProduct(Product $product = null)
     /**
      * Get group name of option by given option type
      *
-     * @param string $type
+     * @param string|null $type
      * @return string
      */
-    public function getGroupByType($type = null)
+    public function getGroupByType($type = null): string
     {
         if ($type === null) {
             $type = $this->getType();
         }
-        $optionGroupsToTypes = [
-            self::OPTION_TYPE_FIELD => self::OPTION_GROUP_TEXT,
-            self::OPTION_TYPE_AREA => self::OPTION_GROUP_TEXT,
-            self::OPTION_TYPE_FILE => self::OPTION_GROUP_FILE,
-            self::OPTION_TYPE_DROP_DOWN => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_RADIO => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_CHECKBOX => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_MULTIPLE => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_DATE => self::OPTION_GROUP_DATE,
-            self::OPTION_TYPE_DATE_TIME => self::OPTION_GROUP_DATE,
-            self::OPTION_TYPE_TIME => self::OPTION_GROUP_DATE,
-        ];
 
-        return $optionGroupsToTypes[$type] ?? '';
+        return $this->typesPool[$type] ?? '';
     }
 
     /**
      * Group model factory
      *
      * @param string $type Option type
-     * @return \Magento\Catalog\Model\Product\Option\Type\DefaultType
+     * @return DefaultType
      * @throws LocalizedException
      */
-    public function groupFactory($type)
+    public function groupFactory($type): DefaultType
```

Comment:
As you can see - earlier it was using all classes from namespace `Magento\Catalog\Model\Product\Option\Type\\` - so potentially not all could be extended from `\Magento\Catalog\Model\Product\Option\Type\DefaultType` - as result it might cause type error here.

Could you double check that?

Reply:
Original class returned DefaultType. It already have too many methods which are required for initialization and rendering custom options in different places. And this DefaultType does not have an interface. It looks like some abstraction, but there no word "Abstract".
I agree with you that we tried to make better for users and at the same time put such restriction. Maybe yes, it will be better to remove strict return, but what to put inside PHPDoc @return? DataObject?